dynamic_instrumentation/src/instrument_memory.rs

```diff
@@ -237,15 +241,50 @@ impl<'a, 'tcx: 'a> CollectFunctionInstrumentationPoints<'a, 'tcx> {
     }
 
     fn should_instrument(&self, local: Local, context: PlaceContext) -> bool {
-        let is_unsafe_ptr = self.body.local_decls[local].ty.is_unsafe_ptr();
+        let is_unsafe_ptr = self.is_shared_region_or_unsafe_ptr(self.body.local_decls[local].ty);
         is_unsafe_ptr && context.is_use()
     }
 
+    fn is_unsafe_ptr(&self, ty: &TyS<'tcx>) -> bool {
+        ty.is_unsafe_ptr()
+    }
+
+    fn is_shared_region_or_unsafe_ptr(&self, ty: &TyS<'tcx>) -> bool {
+        ty.is_unsafe_ptr() || (ty.is_region_ptr() && !ty.is_mutable_ptr())
+    }
+
     fn func_hash(&self) -> mir_loc::DefPathHash {
         self.tcx.def_path_hash(self.body.source.def_id()).convert()
     }
 }
 
+/// Used to strip initital deref from projection sequences
+fn remove_outer_deref<'tcx>(p: &Place<'tcx>, tcx: TyCtxt<'tcx>) -> Place<'tcx> {
+    fn pop_one_projection<'tcx>(p: &Place<'tcx>, tcx: TyCtxt<'tcx>) -> Option<Place<'tcx>> {
+        if p.projection.is_empty() {
+            return None;
+        }
+        Some(Place {
+            local: p.local,
+            projection: tcx.intern_place_elems(&p.projection[..p.projection.len() - 1]),
+        })
+    }
+
+    // For mutable borrows (let _n = &mut place), create a parallel binding that takes
+    // a raw pointer to the same place, without involving _n
+
```

Comment:
This comment doesn't make sense after moving these fns out of their original context.